plugins/inputs/logstash/logstash.go

```diff
@@ -0,0 +1,284 @@
+		return err
+	}
+
+	tags := map[string]string{
+		"node_id": ProcessStats.ID,
+	}
+
+	stats := map[string]interface{}{
+		"process": ProcessStats.Process,
+	}
+
+	for p, s := range stats {
+		f := jsonparser.JSONFlattener{}
+		// parse Json, ignoring strings and bools
+		err := f.FlattenJSON("", s)
+		if err != nil {
+			return err
+		}
+		acc.AddFields("logstash_"+p, f.Fields, tags)
+	}
+
+	return nil
+}
+
+func (l *Logstash) gatherPipelineStats(url string, acc telegraf.Accumulator) error {
+	PipelineStats := &PipelineStats{}
+	if err := l.gatherJSONData(url, PipelineStats); err != nil {
+		return err
+	}
+
+	tags := map[string]string{
+		"node_id": PipelineStats.ID,
+	}
+
+	stats := map[string]interface{}{
+		"events": PipelineStats.Pipeline.Events,
+	}
+
+	// Events
+	for p, s := range stats {
+		f := jsonparser.JSONFlattener{}
+		// parse Json, ignoring strings and bools
+		err := f.FlattenJSON("", s)
+		if err != nil {
+			return err
+		}
+		acc.AddFields("logstash_"+p, f.Fields, tags)
+	}
+
+	// Input Plugins
+	for _, plugin := range PipelineStats.Pipeline.Plugins.Inputs {
+		//plugin := &plugin
+		fields := map[string]interface{}{
+			"queue_push_duration_in_millis": plugin.Events.QueuePushDurationInMillis,
+			"duration_in_millis":            plugin.Events.DurationInMillis,
+			"in":                            plugin.Events.In,
+			"out":                           plugin.Events.Out,
+		}
+		tags := map[string]string{
+			"plugin": plugin.Name,
+			"type":   "input",
+		}
+		acc.AddFields("logstash_plugins", fields, tags)
+	}
+
+	// Filters Plugins
+	for _, plugin := range PipelineStats.Pipeline.Plugins.Filters {
+		//plugin := &plugin
+		fields := map[string]interface{}{
+			"duration_in_millis": plugin.Events.DurationInMillis,
+			"in":                 plugin.Events.In,
+			"out":                plugin.Events.Out,
+		}
+		tags := map[string]string{
+			"plugin": plugin.Name,
+			"type":   "filter",
+		}
+		acc.AddFields("logstash_plugins", fields, tags)
+	}
+
+	// Output Plugins
+	for _, plugin := range PipelineStats.Pipeline.Plugins.Outputs {
+		//plugin := &plugin
+		fields := map[string]interface{}{
+			"duration_in_millis": plugin.Events.DurationInMillis,
+			"in":                 plugin.Events.In,
+			"out":                plugin.Events.Out,
+		}
+		tags := map[string]string{
+			"plugin": plugin.Name,
+			"type":   "output",
+		}
+		acc.AddFields("logstash_plugins", fields, tags)
+	}
+
+	return nil
+}
+
+//Gather is main function to gather all metrics provided by this plugin
+func (l *Logstash) Gather(acc telegraf.Accumulator) error {
+
+	if l.client == nil {
+		client, err := l.createHTTPClient()
+
+		if err != nil {
+			return err
+		}
+		l.client = client
+	}
+
+	jvm_url, err := url.Parse(l.URL + jvmStats)
+	if err != nil {
+		return err
+	}
+	if err := l.gatherJVMStats(jvm_url.String(), acc); err != nil {
+		return err
+	}
+
+	process_url, err := url.Parse(l.URL + processStats)
```

Comment:
Build the urls by making a copy and setting the path:
```
url := jvm_url
url.Path = processStats
```

Consider passing the *url.URL down and changing to string at last second.

Reply:
+1 ^